app/src/app/sections/lia/lia-submission-page/lia-submission-page.component.ts

```diff
@@ -14,31 +16,126 @@ import { LiaHeaderComponent } from './../lia-header';
+  timeWarningDisplayed: boolean;
+  toastMsg: string;
 
-  constructor(private liaService: LiaService,
-              private router: Router,
-              private routeParams: RouteParams) { }
+  constructor(
+    private liaService: LiaService,
+    private router: Router,
+    private routeParams: RouteParams) {
+      this.timeWarningDisplayed = false;
+      this.toastMsg = '';
+  }
 
   ngOnInit() {
     let userId = +this.routeParams.get('userId');
     let liaId = +this.routeParams.get('liaId');
     this.liaService.getUserLia(userId, liaId).then(res => {
       this.lia = res;
-
       if (this.lia.state === 'submitted' || this.lia.state === 'opened') {
         let link = ['LiaLandingPage', { userId: userId, liaId: res.id }];
         this.router.navigate(link);
       }
+
+      if (this.lia.state === 'in_progress') {
+        // Interval that handles progress save on server.
+        this.saveProgressIntervalId = setInterval(() => {
+          let userId = +this.routeParams.get('userId');
+          let lia = this.copyLia(this.lia);
+
+          if (lia.state === 'submitted') {
+            clearTimeout(this.saveProgressIntervalId);
+            return;
+          }
+
+          lia.snippet_code = btoa(lia.submitted_code);
+          lia.submitted_code = '';
+
+          this.liaService.saveProgress(userId, lia);
+        }, 10000);
+      }
+
+      const timeWarning = 60;
+      let remainingSeconds = this.getRemainingSeconds();
+
+      if (remainingSeconds <= 0) {
+        this.doAutoSubmit();
+        return;
+      }
+
+      if (remainingSeconds > timeWarning) {
+        this.timeWarningIntervalId = setInterval(() => {
+          let remainingSeconds = this.getRemainingSeconds();
+
+          if (!this.timeWarningDisplayed && remainingSeconds <= timeWarning) {
+            this.toastMsg = 'You only have ' + timeWarning + ' seconds left to complete the challenge!';
+            this.timeWarningDisplayed = true;
+          } else if (remainingSeconds <= 0) {
+            this.doAutoSubmit();
+            clearTimeout(this.timeWarningIntervalId);
+          }
+        }, 1000);
+      }
     });
   }
 
+  copyLia(src: Lia): Lia {
+    let newLia = {
+      id: src.id,
+      title: src.title,
+      short_description: src.short_description,
+      description: src.description,
+      language: src.language,
+      state: src.state,
+      time: src.time,
+      created_at: src.created_at,
+      username: src.username,
+      user_id: src.user_id,
+      submitted_code: src.submitted_code,
+      snippet_code: src.snippet_code,
+      submitted_at: src.submitted_at,
+      started_at: src.started_at
+    };
+
+    return newLia;
+  }
+
+  doAutoSubmit() {
+    // @TODO: change to a modal dialog
+    alert('Your time has ended. Your progress will be saved now.');
+    this.lia.snippet_code = '';
+    this.lia.submitted_code = btoa(this.lia.submitted_code);
+    this.submitLia(this.lia);
+  }
+
+  getRemainingSeconds(): Number {
```

Comment:
it does, what I meant is to improve this method by passing `lia` object as an argument to make code better organized. it's like: "Hey, `getRemainingSeconds ` method here is `lia` for you, so tell me its remaining seconds"